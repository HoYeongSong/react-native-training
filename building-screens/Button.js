// @flow
import React, {Component} from 'react';
import {
  Platform,
  StyleSheet,
  Text,
  View,
  ActivityIndicator,
  TouchableOpacity,
  Image,
} from 'react-native';

import type {
  ____ViewStyleProp_Internal as ViewStyle,
  ____TextStyleProp_Internal as TextStyle,
  ____ImageStyleProp_Internal as ImageStyle,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

type Styles = {
  container: ViewStyle,
  btn: ViewStyle,
  btnDisabled: ViewStyle,
  txt: TextStyle,
  imgLeft: ImageStyle,
}

const styles: Styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  btn: {
    backgroundColor: 'transparent',
    alignSelf: 'center',
    borderRadius: 4,
    borderWidth: 2,
    width: '100%',
    height: '100%',
    borderColor: 'white',

    alignItems: 'center',
    justifyContent: 'center',
  },
  btnDisabled: {
    backgroundColor: 'rgb(243,243,243)',
    alignSelf: 'center',
    borderRadius: 4,
    borderWidth: 2,
    width: '100%',
    height: '100%',
    borderColor: '#333',

    alignItems: 'center',
    justifyContent: 'center',
  },
  txt: {
    fontSize: 14,
    color: 'white',
  },
  imgLeft: {
    width: 24,
    height: 24,
    position: 'absolute',
    left: 16,
  },
});

type Props = {
  containerStyle?: ViewStyle,
  disabledStyle?: ViewStyle,
  isLoading?: boolean,
  isDisabled?: boolean,
  style?: ViewStyle,
  textStyle?: TextStyle,
  imgLeftStyle?: ImageStyle,
  indicatorColor?: string,
  children?: string,
  onPress?: () => void,
  imgLeftSrc?: any,
};

export default class Shared extends Component<Props> {
  static defaultProps = {
    containerStyle: styles.container,
    isLoading: false,
    isDisabled: false,
    style: styles.btn,
    textStyle: styles.txt,
    imgLeftStyle: styles.imgLeft,
    indicatorColor: 'white',
    disabledStyle: styles.btnDisabled,
  }

  render() {
    if (this.props.isDisabled) {
      return (
        <View style={this.props.containerStyle}>
          <View style={this.props.disabledStyle}>
            <Text style={this.props.textStyle}>{this.props.children}</Text>
          </View>
        </View>
      )
    }
    if (this.props.isLoading) {
      return (
        <View style={this.props.containerStyle}>
          <View style={this.props.style}>
            <ActivityIndicator size='small' color={this.props.indicatorColor}/>
          </View>
        </View>
      )
    }
    return (
      <View style={this.props.containerStyle}>
        <TouchableOpacity
          onPress={this.props.onPress}
        >
          <View style={this.props.style}>
            {
              this.props.imgLeftStyle
                ? <Image
                  style={this.props.imgLeftStyle}
                  source={this.props.imgLeftSrc}
                />
                : <View />
            }
            <Text style={this.props.textStyle}>{this.props.children}</Text>
          </View>
        </TouchableOpacity>
      </View>
    );
  }
}
