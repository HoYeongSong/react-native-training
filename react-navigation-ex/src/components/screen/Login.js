// @flow
import React, {Component} from 'react';
import {
  StyleSheet,
  Image,
  Text,
  TouchableOpacity,
  View,
  ScrollView,
} from 'react-native';

import type {
  ____ViewStyleProp_Internal as ViewStyle,
  ____TextStyleProp_Internal as TextStyle,
  ____ImageStyleProp_Internal as ImageStyle,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import TextInput from '../shared/TextInput';
import Button from '../shared/Button';

import { IC_APP } from '../../utils/Icons';
import { colors } from '../../utils/Styles';

type Style = {
  container: ViewStyle,
  iconWrapper: ViewStyle,
  icon: ImageStyle,
  iconTxt: TextStyle,
  wrapper: ViewStyle,
  viewBtnWrapper: ViewStyle,
  btnSignup: ViewStyle,
  txtSignUp: TextStyle,
  btnLogin: ViewStyle,
  txtLogin: TextStyle,
  touchForgotPw: ViewStyle,
  txtForgotPw: TextStyle,
  txtCopyright: TextStyle,
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'flex-start',
    alignSelf: 'stretch',
  },
  iconWrapper: {
    position: 'absolute',
    top: 76,
    left: 40,

    flexDirection: 'column',
    alignItems: 'flex-start',
  },
  icon: {
    width: 60,
    height: 48,
  },
  iconTxt: {
    fontSize: 20,
    color: colors.dusk,
    marginTop: 16,
    fontWeight: 'bold',
  },
  wrapper: {
    marginTop: 260,
    width: '78%',
    height: 300,
    alignSelf: 'center',

    flexDirection: 'column',
    alignItems: 'center',
  },
  wrapperInput: {
    alignSelf: 'stretch',
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
  },
  input: {
    alignSelf: 'stretch',
    color: colors.dusk,
    fontSize: 16,
    paddingVertical: 22,
    paddingHorizontal: 20,
    borderWidth: 1,
    borderColor: colors.paleGray,
  },
  viewBtnWrapper: {
    alignSelf: 'stretch',
    marginTop: 20,
    height: 60,

    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  btnSignup: {
    backgroundColor: 'transparent',
    borderRadius: 4,
    borderWidth: 1,
    height: '100%',
    width: '100%',
    borderColor: colors.dodgerBlue,

    alignItems: 'center',
    justifyContent: 'center',
  },
  txtSignup: {
    fontSize: 16,
    fontWeight: 'bold',
    color: colors.dodgerBlue,
  },
  btnLogin: {
    backgroundColor: colors.dodgerBlue,
    borderColor: colors.dodgerBlue,
    borderRadius: 4,
    borderWidth: 1,
    height: 60,
    shadowColor: colors.dodgerBlue,
    shadowOffset: {
      width: 0,
      height: 10,
    },
    shadowRadius: 4,
    shadowOpacity: 0.3,


    alignItems: 'center',
    justifyContent: 'center',
  },
  txtLogin: {
    fontSize: 16,
    fontWeight: 'bold',
    color: 'white',
  },
  touchForgotPw: {
    marginTop: 20,
  },
  txtForgotPw: {
    fontSize: 12,
    color: colors.dodgerBlue,
    textDecorationLine: 'underline',
  },
  txtCopyright: {
    marginTop: 80,
    fontSize: 12,
    color: colors.cloudyBlue,
  },
});

type Props = {
  navigation: any,
};
type State = {
  email: string,
  password: string,
  isLoggingin: boolean,
};

export default class App extends Component<Props, State> {
  static navigationOptions = {
    headerStyle: {
      borderBottomWidth: 0,
    },
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      isLoggingin: false,
    };
  }
  render() {
    return (
      <ScrollView>
        <View style={styles.container}>
          <View style={styles.iconWrapper}>
            <Image style={styles.icon} source={ IC_APP }/>
            <Text style={styles.iconTxt}>Hello</Text>
          </View>
          <View style={styles.wrapper}>
            <View style={styles.wrapperInput}>
              <TextInput
                style={{ marginTop: 60 }}
                txt={this.state.email}
                txtHint='Please write email address.'
                placeHolderTextColor={colors.blueyGray}
                onTextChanged={(text) => {
                  this.onTextChanged('EMAIL', text);
                }}
              />
              <TextInput
                style={{ marginTop: 8 }}
                txt={this.state.password}
                txtHint='Please write password.'
                placeHolderTextColor={colors.blueyGray}
                onTextChanged={(text) => {
                  this.onTextChanged('PASSWORD', text);
                }}
                isPassword
              />
            </View>
            <View style={styles.viewBtnWrapper}>
              <Button
                containerStyle={{ flex: 1 }}
                onPress={() => {
                  this.props.navigation.navigate('Signup');
                }}
                style={styles.btnSignup}
                textStyle={styles.txtSignup}
              >Sign Up</Button>
              <View style={{ width: 8 }}/>
              <Button
                containerStyle={{ flex: 1 }}
                isLoading={this.state.isLoggingin}
                onPress={() => {
                  this.setState({
                    isLoggingin: true
                  }, () => {
                    setTimeout(() => {
                      this.setState({
                        isLoggingin: false,
                      });
                    }, 3000);
                  })
                }}
                style={styles.btnLogin}
                textStyle={styles.txtLogin}
              >Login</Button>
            </View>
            <TouchableOpacity
              style={styles.touchForgotPw}
            >
              <Text style={styles.txtForgotPw}>Forgot password?</Text>
            </TouchableOpacity>
            <Text style={styles.txtCopyright}>copyright by dooboolab.com</Text>
          </View>
        </View>
      </ScrollView>
    );
  }

  onTextChanged = (type: string, text: string) => {
    switch(type) {
      case 'EMAIL':
        this.setState({
          email: text,
        });
      break;
      case 'PASSWORD':
        this.setState({
          password: text,
        });
      break;
    }
  }
}
