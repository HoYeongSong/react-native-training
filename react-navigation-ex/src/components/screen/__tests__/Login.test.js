import 'react-native';
import React from 'react';
import Login, { styles } from '../Login';

import TextInput from '../../shared/TextInput';
import Button from '../../shared/Button';

import renderer from 'react-test-renderer';
import { shallow, render } from 'enzyme';

let props;
let wrapper;
let tree;
let instance;

describe('rendering test', () => {
  beforeEach(() => {
    props = {
      navigation: {
        navigate: jest.fn(),
      },
    };
    wrapper = shallow(<Login {...props}/>);
    tree = renderer.create(<Login {...props} />).toJSON();
    instance = renderer.create(<Login {...props}/>).getInstance();
    // instance = wrapper.instance();
  });

  it('should renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should renders inital state', () => {
    expect(instance.state.email).toEqual('');
    expect(instance.state.password).toEqual('');
    expect(instance.state.isLoggingIn).toEqual(false);
  });

  it('should find components', () => {
    expect(wrapper.find('ScrollView')).toHaveLength(1);
    expect(wrapper.find('Text')).toHaveLength(3);
    expect(wrapper.find(TextInput)).toHaveLength(2);
    expect(wrapper.find(Button)).toHaveLength(2);
    expect(wrapper.find('Text').at(0)).toHaveLength(1);
  });

  it('should have default style', () => {
    expect(wrapper.find('View').at(0).prop('style')).toEqual(styles.container);
    expect(wrapper.find('View').at(0).childAt(0).prop('style')).toEqual(styles.iconWrapper);
    expect(wrapper.find('View').at(0).childAt(1).prop('style')).toEqual(styles.wrapper);
  });

  it('should change state onTextChanged', () => {
    instance.onTextChanged('EMAIL', 'aa@aa.aa');
    expect(instance.state.email).toEqual('aa@aa.aa');
    instance.onTextChanged('PASSWORD', 'aaaaaa');
    expect(instance.state.password).toEqual('aaaaaa');
  });
});

describe('interaction', () => {
  beforeEach(() => {
    props = {
      navigation: {
        navigate: jest.fn(),
      },
    };
    wrapper = shallow(<Login {...props} />);
    tree = renderer.create(<Login {...props} />).toJSON();
    instance = renderer.create(<Login {...props}/>).getInstance();
    // instance = wrapper.instance();
  });

  it('should call onLogin callback', () => {
    jest.useFakeTimers();
    const spy = jest.spyOn(wrapper.instance(), 'onLogin');
    const loginBtn = wrapper.find('#login');
    loginBtn.props().onPress();
    expect(spy).toHaveBeenCalled();
    expect(wrapper.instance().state.isLoggingIn).toEqual(true);
    jest.runAllTimers();
    expect(wrapper.instance().state.isLoggingIn).toEqual(false);
  });

  it('should go to FindPw on goToForgotPw callback', () => {
    const findPwBtn = wrapper.find('TouchableOpacity');
    findPwBtn.props().onPress();
    expect(props.navigation.navigate).toHaveBeenCalledWith('FindPw');
  });
});