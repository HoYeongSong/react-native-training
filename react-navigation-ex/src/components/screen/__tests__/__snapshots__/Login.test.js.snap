// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering test should renders as expected 1`] = `
<ScrollView>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "alignSelf": "stretch",
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flexDirection": "column",
          "left": 40,
          "position": "absolute",
          "top": 76,
        }
      }
    >
      <Component
        source={
          Object {
            "testUri": "../../../assets/icons/symbol.png",
          }
        }
        style={
          Object {
            "height": 48,
            "width": 60,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "rgb(65,77,107)",
            "fontSize": 20,
            "fontWeight": "bold",
            "marginTop": 16,
          }
        }
      >
        Hello
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "flexDirection": "column",
          "height": 300,
          "marginTop": 260,
          "width": "78%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "stretch",
            "flexDirection": "column",
            "justifyContent": "flex-start",
          }
        }
      >
        <Shared
          labelStyle={
            Object {
              "color": "rgb(134,154,183)",
              "fontSize": 12,
              "marginBottom": 8,
            }
          }
          labelStyleFocus={
            Object {
              "color": "rgb(58,139,255)",
              "fontSize": 12,
              "marginBottom": 8,
            }
          }
          onTextChanged={[Function]}
          placeHolderTextColor="rgb(134,154,183)"
          style={
            Object {
              "marginTop": 60,
            }
          }
          txt=""
          txtHint="Please write email address."
        />
        <Shared
          isPassword={true}
          labelStyle={
            Object {
              "color": "rgb(134,154,183)",
              "fontSize": 12,
              "marginBottom": 8,
            }
          }
          labelStyleFocus={
            Object {
              "color": "rgb(58,139,255)",
              "fontSize": 12,
              "marginBottom": 8,
            }
          }
          onTextChanged={[Function]}
          placeHolderTextColor="rgb(134,154,183)"
          style={
            Object {
              "marginTop": 8,
            }
          }
          txt=""
          txtHint="Please write password."
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "space-between",
            "marginTop": 20,
          }
        }
      >
        <Shared
          containerStyle={
            Object {
              "flex": 1,
            }
          }
          disabledStyle={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "rgb(243,243,243)",
              "borderColor": "#333",
              "borderRadius": 4,
              "borderWidth": 2,
              "height": "100%",
              "justifyContent": "center",
              "width": "100%",
            }
          }
          imgLeftStyle={
            Object {
              "height": 24,
              "left": 16,
              "position": "absolute",
              "width": 24,
            }
          }
          indicatorColor="white"
          isDisabled={false}
          isLoading={false}
          onPress={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "rgb(58,139,255)",
              "borderRadius": 4,
              "borderWidth": 1,
              "height": "100%",
              "justifyContent": "center",
              "width": "100%",
            }
          }
          textStyle={
            Object {
              "color": "rgb(58,139,255)",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          Sign Up
        </Shared>
        <View
          style={
            Object {
              "width": 8,
            }
          }
        />
        <Shared
          containerStyle={
            Object {
              "flex": 1,
            }
          }
          disabledStyle={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "rgb(243,243,243)",
              "borderColor": "#333",
              "borderRadius": 4,
              "borderWidth": 2,
              "height": "100%",
              "justifyContent": "center",
              "width": "100%",
            }
          }
          id="login"
          imgLeftStyle={
            Object {
              "height": 24,
              "left": 16,
              "position": "absolute",
              "width": 24,
            }
          }
          indicatorColor="white"
          isDisabled={false}
          isLoading={false}
          onPress={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgb(58,139,255)",
              "borderColor": "rgb(58,139,255)",
              "borderRadius": 4,
              "borderWidth": 1,
              "height": 60,
              "justifyContent": "center",
              "shadowColor": "rgb(58,139,255)",
              "shadowOffset": Object {
                "height": 10,
                "width": 0,
              },
              "shadowOpacity": 0.3,
              "shadowRadius": 4,
            }
          }
          textStyle={
            Object {
              "color": "white",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          Login
        </Shared>
      </View>
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "marginTop": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgb(58,139,255)",
              "fontSize": 12,
              "textDecorationLine": "underline",
            }
          }
        >
          Forgot password?
        </Text>
      </TouchableOpacity>
      <Text
        style={
          Object {
            "color": "rgb(175,194,219)",
            "fontSize": 12,
            "marginTop": 80,
          }
        }
      >
        copyright by dooboolab.com
      </Text>
    </View>
  </View>
</ScrollView>
`;
